File: src/main/java/com/example/gpiApp/service/SavedFilterService.java
package com.example.gpiApp.service;

import com.example.gpiApp.entity.SavedFilter;
import com.example.gpiApp.repository.SavedFilterRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * CRUD for server-side saved filters. Tenant-scoped by the @Filter (active in these @Transactional
 * methods); a filter can only be edited or deleted by its owner. Listing returns the caller's own
 * filters plus any a colleague shared org-wide.
 */
@Service
@RequiredArgsConstructor
public class SavedFilterService {

    private final SavedFilterRepository repository;

    @Transactional(readOnly = true)
    public List<SavedFilter> list(Long userId, String resource) {
        return repository.findVisible(userId, resource == null || resource.isBlank() ? "tasks" : resource);
    }

    @Transactional
    public SavedFilter create(SavedFilter f, Long userId) {
        if (f.getName() == null || f.getName().isBlank()) throw new IllegalArgumentException("Filter name is required.");
        f.setId(null);
        f.setUserId(userId);
        if (f.getResource() == null || f.getResource().isBlank()) f.setResource("tasks");
        return repository.save(f); // TenantListener stamps the org
    }

    @Transactional
    public SavedFilter update(Long id, SavedFilter patch, Long userId) {
        SavedFilter f = getOwned(id, userId);
        if (patch.getName() != null && !patch.getName().isBlank()) f.setName(patch.getName());
        if (patch.getCriteria() != null) f.setCriteria(patch.getCriteria());
        f.setShared(patch.isShared());
        return repository.save(f);
    }

    @Transactional
    public void delete(Long id, Long userId) {
        repository.delete(getOwned(id, userId));
    }

    private SavedFilter getOwned(Long id, Long userId) {
        SavedFilter f = repository.findById(id).orElseThrow(() -> new AccessDeniedException("Filter not found"));
        if (!userId.equals(f.getUserId())) throw new AccessDeniedException("This filter belongs to another user.");
        return f;
    }
}

File: frontend/src/app/core/services/saved-filter.service.ts
import { Injectable } from '@angular/core';
import { HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';

export interface SavedFilter {
  id?: number;
  name: string;
  resource?: string;
  criteria?: string | null;   // opaque JSON string (frontend-defined)
  shared?: boolean;
  createdAt?: string;
}

/** Server-side saved filters / views (per-user, tenant-scoped, optionally shared org-wide). */
@Injectable({ providedIn: 'root' })
export class SavedFilterService {
  private base = '/saved-filters';
  constructor(private api: ApiService) {}

  list(resource = 'tasks'): Observable<SavedFilter[]> {
    return this.api.get<SavedFilter[]>(this.base, new HttpParams().set('resource', resource));
  }
  create(f: SavedFilter): Observable<SavedFilter> { return this.api.post<SavedFilter>(this.base, f); }
  update(id: number, f: SavedFilter): Observable<SavedFilter> { return this.api.put<SavedFilter>(`${this.base}/${id}`, f); }
  delete(id: number): Observable<any> { return this.api.delete<any>(`${this.base}/${id}`); }
}
